### PR TITLE
Use selection.splitNode() to preserve tags while splitting sentences

### LINKED ISSUE #3455: [Mobile][Web] Split thought with color splits the html hardcode
**Step to repro**

1. Create a new thought
2. Add a any collor to the thought
3. Tap or click Split button on tthe toolbar
4. Observe the splitted thought

**Current behavior:**
Split thought is showing html hardcode 

**Expected:**
Splitting should not be implemented on the hardcode.

Image

Image

### PR DESCRIPTION
**Pull request by ethan-james**
Fixes #3455 and #3615

#3455 has a bit of a different root cause than #3615, but I needed to alter the comma-splitting logic in different ways for both, so it would cause merge conflicts to separate them.

#3455 had two issues: semicolons in style tags, and commas in style tags. Semicolon-splitting could be solved by adding an `isStyle` function that performs a similar check to the existing `isUrl` function. Comma-splitting logic happened in a later step, and the logic wasn't as robust. So I added a `separateByComma` function that does the following:

1. replaces commas within style blocks with a comma token
2. replaces commas outside of style blocks with the separator token
3. replaces the comma token with a `,`

For #3615, the solution was to place the markup into a div and use `selection.splitNode()` to preserve tags when splitting into sentences. It was necessary to traverse nodes using `offsetFromClosestParent`, but the end result is that the text is extracted from the div with all tags correctly closed using native browser behavior.

### CODE REVIEW

File: src/commands/__tests__/splitSentences.ts [hunk 1]
```diff
@@ -30,6 +32,64 @@ describe('splitSentences', () => {
...
+    executeCommand(splitSentencesCommand, { store })
+
+    const exported = exportContext(store.getState(), [HOME_TOKEN], 'text/plain')
+    expect(exported).toBe(`- __ROOT__
+  - **This is sentence one.**
+  - **This is sentence two.**
+  - **This is sentence three.**`)
+  })
+
+  it('splits a thought with a single sentence plus font tags', () => {
```
Comment (raineorshine) [inline src/commands/__tests__/splitSentences.ts]: There is no split in this case, so we can say: `does not split a thought with a single sentence and font tag`

File: src/commands/__tests__/splitSentences.ts [hunk 2]
```diff
@@ -30,6 +32,64 @@ describe('splitSentences', () => {
   - This is sentence three.`)
   })
 
+  it('splits a thought with multiple sentences and styles', () => {
```
Comment (raineorshine) [inline src/commands/__tests__/splitSentences.ts]: "and styles" is too general, Let's say "multiple sentences in a bold tag".

We want to keep the tests and test descriptions as concrete as possible and not make assumptions about other styles, which may or may not be handled differently.

File: src/commands/__tests__/splitSentences.ts [hunk 3]
```diff
@@ -30,6 +32,64 @@ describe('splitSentences', () => {
...
+    expect(exported).toBe(`<ul>
+  <li>__ROOT__  
+    <ul>
+      <li><font color="#000000" style="background-color: rgb(0, 214, 136);">font</font></li>
+    </ul>
+  </li>
+</ul>`)
+  })
+
+  it('splits a thought with a multiple comma-delimited sentences plus font tags', () => {
```
Comment (raineorshine) [inline src/commands/__tests__/splitSentences.ts]: This says font "tags" but the test case only contains one font tag.

**Comment by ethan-james** [comment]
> Thanks for the submission and the clear PR description. This looks fine to me.
> 
> Just a few recommendations to improve the test titles.
> 
> Since `splitNode` is now HTML aware, it may have made sense to migrate to proper HTML parsing and split nodes at the AST level before serializing it back into an HTML string. The risk is that we encounter additional edge cases that this algorithm can't handle, or that the complexity of the regex approach continues to grow, in which case we'll have to throw away a lot of this work. However, since we are already herIe I'm okay to keep what we have and proceed tentatively.

I did consider that, and then I decided it might be a case of the cure being worse than the disease. We may need to go that way in the future, but I identified that:

1.  there are only 5 markup tags that we're using (that I know of)
2.  font tag is the only one with an attribute, and
3. the style attribute is the only attribute

It's also possible for `.` or `!` to appear in style rules, but I don't think it's possible for us to generate such rules. Let me know if there's something else I need to test.

**Review by raineorshine — APPROVED** [review]
Good, I'm glad you've considered it. We'll see how this holds up. If we don't encounter more edge cases then I'm happy to call it a win and move on.

> I did consider that, and then I decided it might be a case of the cure being worse than the disease. We may need to go that way in the future, but I identified that:
> 
> 1. there are only 5 markup tags that we're using (that I know of)
> 2. font tag is the only one with an attribute, and
> 3. the style attribute is the only attribute

Yes, I think that's about right. My worry is with multiple overlapping styles with different ranges, but I haven't stress tested it.

@BayuAri Feel free to do another round of UX stress testing of split sentences on `main` to uncover any faults! (I merged this prematurely.)
